src/libraries/System.Runtime.Numerics/src/System/Numerics/BigNumber.cs

```diff
@@ -380,79 +382,254 @@ internal static BigInteger ParseBigInteger(ReadOnlySpan<char> value, NumberStyle
             if (!TryValidateParseStyleInteger(style, out e))
                 throw e;
 
-            BigInteger result = BigInteger.Zero;
-            if (!TryParseBigInteger(value, style, info, out result))
+            if (!TryParseBigInteger(value, style, info, out BigInteger result))
             {
                 throw new FormatException(SR.Overflow_ParseBigInteger);
             }
             return result;
         }
 
-        private static unsafe bool HexNumberToBigInteger(ref BigNumberBuffer number, ref BigInteger value)
+        private static bool HexNumberToBigInteger(ref BigNumberBuffer number, out BigInteger result)
         {
             if (number.digits == null || number.digits.Length == 0)
+            {
+                result = default;
                 return false;
+            }
+
+            const int DigitsPerBlock = 8;
+
+            int totalDigitCount = number.digits.Length - 1;   // Ignore trailing '\0'
+            int blockCount, partialDigitCount;
 
-            int len = number.digits.Length - 1; // Ignore trailing '\0'
-            byte[] bits = new byte[(len / 2) + (len % 2)];
+            if ((uint)totalDigitCount % DigitsPerBlock == 0)
+            {
+                blockCount = totalDigitCount / DigitsPerBlock;
+                partialDigitCount = 0;
+            }
+            else
+            {
+                blockCount = Math.DivRem(totalDigitCount, DigitsPerBlock, out int mod) + 1;
+                partialDigitCount = DigitsPerBlock - mod;
+            }
```

Comment:
We compute the quotient and remainder in both paths, so we could simplify to something like:
```suggestion
            blockCount = Math.DivRem(totalDigitCount, DigitsPerBlock, out int remainder);

            if (remainder == 0)
            {
                partialDigitCount = 0;
            }
            else
            {
                blockCount += 1;
                partialDigitCount = DigitsPerBlock - remainder;
            }
```

Reply:
We might also be able to just have `blockCount` and `partialDigitCount` be unsigned most of the way through, at least until they actually need to be signed again.